homeassistant/components/mcp23017/__init__.py

```diff
@@ -1,3 +1,372 @@
+
+    # Unload related platform
+    await hass.config_entries.async_forward_entry_unload(
+        config_entry, config_entry.data[CONF_FLOW_PLATFORM]
+    )
+
+    i2c_address = config_entry.data[CONF_I2C_ADDRESS]
+
+    # DOMAIN data async mutex
+    async with MCP23017_DATA_LOCK:
+        if i2c_address in hass.data[DOMAIN]:
+            component = hass.data[DOMAIN][i2c_address]
+
+            # Unlink entity from component
+            await hass.async_add_executor_job(
+                functools.partial(
+                    component.unregister_entity, config_entry.data[CONF_FLOW_PIN_NUMBER]
+                )
+            )
+
+            # Free component if not linked to any entities
+            if component.has_no_entities:
+                await hass.async_add_executor_job(component.stop_polling)
+                hass.data[DOMAIN].pop(i2c_address)
+
+                _LOGGER.info(
+                    "%s@0x%02x component destroyed",
+                    type(component).__name__,
+                    i2c_address,
+                )
+
+    return True
+
+
+async def async_get_or_create(hass, config_entry, entity):
+    """Get or create a MCP23017 component from entity i2c address."""
+
+    i2c_address = entity.address
+
+    # DOMAIN data async mutex
+    try:
+        async with MCP23017_DATA_LOCK:
+            if i2c_address in hass.data[DOMAIN]:
+                component = hass.data[DOMAIN][i2c_address]
+            else:
+                # Try to create component when it doesn't exist
+                component = await hass.async_add_executor_job(
+                    functools.partial(MCP23017, DEFAULT_I2C_BUS, i2c_address)
+                )
+                hass.data[DOMAIN][i2c_address] = component
+
+                # Start polling thread if hass is already running
+                if hass.is_running:
+                    component.start_polling()
+
+                # Register a device combining all related entities
+                devices = await device_registry.async_get_registry(hass)
+                devices.async_get_or_create(
+                    config_entry_id=config_entry.entry_id,
+                    identifiers={(DOMAIN, i2c_address)},
+                    manufacturer="MicroChip",
+                    model=DOMAIN,
+                    name=f"{DOMAIN}@0x{i2c_address:02x}",
+                )
+
+            # Link entity to component
+            await hass.async_add_executor_job(
+                functools.partial(component.register_entity, entity)
+            )
+
+    except ValueError as error:
+        component = None
+        await hass.config_entries.async_remove(config_entry.entry_id)
+
+        hass.components.persistent_notification.create(
+            f"Error: Unable to access {DOMAIN}-0x{i2c_address:02x} ({error})",
+            title=f"{DOMAIN} Configuration",
+            notification_id=f"{DOMAIN} notification",
+        )
+
+    return component
+
+
+class MCP23017(threading.Thread):
```

Comment:
This class looks too low level for inclusion in Home Assistant directly. We want to use a library for device integration code.

Reply:
@MartinHjelmare I do agree we you in general but in this case, it would basically mean to remove 7 lines (dict setter/getter L220-L227) and replace statements like:

`self[GPIOA] = value`

by:

`self._device.gpioa(value)`

Everything else would stay the same, i.e.  comparable readability but no dependencies to foreign libraires (adafruit-circuitpython-mcp230xx, Adafruit-Blinka, adafruit-circuitpython-busdevice and assosciated maintenance issues like #31719, #40192) that are not abstracting anything or adding value in this context.